torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py

```diff
@@ -0,0 +1,51 @@
+import torch
+
+from .base_sparsifier import BaseSparsifier
+
+
+class NearlyDiagonalSparsifier(BaseSparsifier):
+    r"""Nearly Diagonal Sparsifier
+
+    This sparsifier creates a nearly diagonal mask to be applied to the weight matrix. 
+    Nearly Diagonal Matrix is a matrix that contains non-zero elements near the diagonal and the rest are zero. 
+    An example of a nearly diagonal matrix with degree (or nearliness) 3 and 5 are follows respectively.
+    1 1 0 0       1 1 1 0
+    0 1 1 0       1 1 1 1
+    0 1 1 1       1 1 1 1
+    0 0 1 1       0 1 1 1
+    Note that a nearly diagonal matrix with degree 1 is just a matrix with main diagonal populated
+
+    This sparsifier is controlled by one variables:
+    1. `nearliness` defines the number of non-zero diagonal lines that are closest to the main diagonal. 
+        Currently - supports only odd number
+
+    Args:
+        nearliness: The degree of nearliness
+
+    """
+    def __init__(self, nearliness: int = 1):
+        defaults = {'nearliness': nearliness}
+        super().__init__(defaults=defaults)
+
+    def update_mask(self, layer, nearliness: int,
+                    **kwargs):
+        mask = layer.parametrizations.weight[0].mask
+        mask.data = torch.zeros_like(mask)
+        if nearliness <= 0:
+            return
+
+        weight = layer.weight
+        height, width = weight.shape
+
+        if nearliness % 2 == 0:
+            raise ValueError("nearliness can only be an odd number")
+        dist_to_diagonal = nearliness // 2
+        # check
+        if dist_to_diagonal >= min(height, width):
+            raise ValueError("nearliness cannot be larger than the "
+                            "dimensions of weight matrix.")
+        # create mask
+        for row in range(0, height):
```

Comment:
just curious, can this be done in a vectorized way? (maybe a good moment to introduce a toeplitz matrix constructor? some options from numpy at https://stackoverflow.com/questions/52463972/generating-banded-matrices-using-numpy)

Reply:
Thanks for the suggestion. We could do this by identifying the entries that needs to be set and just fill those entries.  Have implemented this. But not sure how we could use a Toeplitz matrix to accomplish this.

Reply:
Correct

Reply:
So a toeplitz matrix or banded matrix constructor should be helpful for this

Reply:
@vadimkantorov Because there is no toeplitz constructor kernel implemented, do you think it's fine to put it as a comment in the docstring, and leave it as is?

Reply:
yep, I think so! then if it's implemented, this code can be updated

Reply:
Added the comment in doc-string. Will update once implemented.